### ISSUE TITLE
Dependencies of component providers aren't downloaded

### ISSUE THREAD

**Issue by cnunciato** [pulumi/pulumi#10521]
Given the following YAML program, which uses the Synced Folder component (a package built from [pulumi-component-provider-ts-boilerplate](https://github.com/pulumi/pulumi-component-provider-ts-boilerplate)):

```
name: yaml-synced-folder-test
runtime: yaml

resources:
  bucket:
    type: aws:s3:Bucket

  bucket-folder:
    type: synced-folder:index:S3BucketFolder
    properties:
      path: ./site   # You'll need a folder with some files here.
      bucketName: ${bucket.bucket}
```

... the initial update fails because the provider's resource plugin wasn't downloaded:

```
pulumi up
Previewing update (dev)

View Live: https://app.pulumi.com/cnunciato/yaml-synced-folder-test/dev/previews/bbb32df0-7f6d-4640-97f7-ed34baa7b9d6

Downloading plugin: 38.06 MiB / 38.06 MiB [=========================] 100.00% 4s
[resource plugin synced-folder-0.0.8] installing
Downloading plugin: 126.54 MiB / 126.54 MiB [=======================] 100.00% 9s
[resource plugin aws-5.13.0] installing

     Type                                   Name                         Plan       Info
 +   pulumi:pulumi:Stack                    yaml-synced-folder-test-dev  create     
 +   ├─ aws:s3:Bucket                       bucket                       create     
 +   ├─ synced-folder:index:S3BucketFolder  bucket-folder                create     
     └─ pulumi:providers:aws                default_5_12_1                          1 error
 
Diagnostics:
  pulumi:providers:aws (default_5_12_1):
    error: no resource plugin 'pulumi-resource-aws' found in the workspace at version v5.12.1 or on your $PATH, install the plugin using `pulumi plugin install resource aws v5.12.1`
```

This is presumably happening because the current version of pulumi-aws is `5.13` (which does get downloaded), whereas the latest synced-folder was built against `pulumi-aws@5.12.1`. The following is a snippet from my local `yarn.lock` that shows this:

```
"@pulumi/aws@^5.0.0":
  version "5.12.1"
  resolved "https://registry.yarnpkg.com/@pulumi/aws/-/aws-5.12.1.tgz#6fdae06c825e852392fc63f4dbe835626acbe36f"
  integrity sha512-QRuqXgNnDNWQZu0xcGHhbgLLF+dYyd9wt2dfsIy5UeP8nExaWmWv0DS4AhWO9KaaUGH+ifqsmX7RdfAzDmVBPw==
  dependencies:
    "@pulumi/pulumi" "^3.0.0"
    aws-sdk "^2.0.0"
    builtin-modules "3.0.0"
    mime "^2.0.0"
    read-package-tree "^5.2.1"
    resolve "^1.7.1"
```

To repro, ensure you don't have `5.12.1` installed already, then run `pulumi up`. I would expect Pulumi to download 5.12.1 and for the update to work without requiring an additional manual install.

**Comment by t0yv0** [comment]
CC @AaronFriel 

Logging discussion with Friel we had async on this. It would be better UX to kick off automatic plugin installation at the point where we fail here, fixing the primary user issue.  IN the long term we may need to think about Pulumi lock files and dep management tools like `pulumi why`. The runtime dependency (hidden) of a component implementation here needs a different AWS Provider version from the SDK used in the program, and while that should be OK it can be surprising at least in terms of dependency bloat, and having tools eventually that explain this would be great.

**Comment by Frassle** [comment]
I think this might just be https://github.com/pulumi/pulumi/issues/5619?

**Comment by t0yv0** [comment]
Hmm yes! The lockfile ideas can go into the extension of #5619 ; but to be clear this particular bug is asking to extend the magic auto-install behavior to cover the case of a runtime dependency of the synced-folder provider.

**Comment by Frassle** [comment]
> but to be clear this particular bug is asking to extend the magic auto-install behavior to cover the case of a runtime dependency of the synced-folder provider.

I think the easiest fix here is to just download missing plugins as we hit them. That will close #5619 off, and unblock this use case, not ideally but at least not p1 bad. 

We can then leave this ticket to track adding plugin lookup from providers, not just programs. Which would be good (and I like your other ideas of lockfiles, `pulumi why`, and no auto downloads) but that's much more work but at least won't need doing so urgently with #5619 done.

**Comment by t0yv0** [comment]
Yes! Sorry bad protocol on my part. Did not mean to ask for increased scope. Should've logged the ideas as separate ticket.

**Comment by cnunciato** [comment]
I believe this is fixed -- is that right? Or is there more y'all want to do?

**Comment by AaronFriel** [comment]
I think this was reopened due to a revert in 3.39.3.
